src/main/java/org/beanpod/switchboard/controller/DeviceController.java

```diff
@@ -10,65 +11,85 @@
-    }
-    return ResponseEntity.ok(device.get());
+  @Override
+  public ResponseEntity<DeviceModel> retrieveDevice(@PathVariable String serialNumber) {
+    DeviceDTO deviceDto =
+        service
+            .findDevice(serialNumber)
+            .orElseThrow(() -> new ExceptionType.DeviceNotFoundException(serialNumber));
+
+    return Optional.of(deviceDto)
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(this::getUnknownException);
   }
 
-  @PostMapping
-  public ResponseEntity<DeviceDTO> createDevice(@RequestBody @Valid DeviceDTO device) {
-    Optional<DeviceDTO> deviceLookup = service.findDevice(device.getSerialNumber());
+  @Override
+  public ResponseEntity<DeviceModel> createDevice(@Valid CreateDeviceRequest createDeviceRequest) {
+    Optional<DeviceDTO> deviceLookup = service.findDevice(createDeviceRequest.getSerialNumber());
     if (deviceLookup.isPresent()) {
-      throw new ExceptionType.DeviceAlreadyExistsException(device.getSerialNumber());
+      throw new ExceptionType.DeviceAlreadyExistsException(createDeviceRequest.getSerialNumber());
     }
-    return ResponseEntity.ok(service.save(device));
+
+    return Optional.of(createDeviceRequest)
+        .map(createRequest -> service.createDevice(createRequest, request.getRemoteAddr()))
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(this::getUnknownException);
   }
 
-  @DeleteMapping("/{serialNumber}")
+  @Override
   @Transactional
   public ResponseEntity<String> deleteDevice(@PathVariable String serialNumber) {
     Long response = service.deleteDevice(serialNumber);
     if (response != 1) {
       throw new ExceptionType.DeviceNotFoundException(serialNumber);
     }
-    return ResponseEntity.ok("Device with serial number " + serialNumber + " Deleted");
+    return ResponseEntity.ok("Device with serial number " + serialNumber + " deleted");
   }
 
-  @PutMapping
+  @Override
   @Transactional
-  public ResponseEntity<DeviceDTO> updateDevice(@RequestBody @Valid DeviceDTO device) {
-    Optional<DeviceDTO> deviceLookup = service.findDevice(device.getSerialNumber());
-    if (deviceLookup.isEmpty()) {
-      throw new ExceptionType.DeviceNotFoundException(device.getSerialNumber());
-    }
-    return ResponseEntity.ok(service.save(device));
+  public ResponseEntity<DeviceModel> updateDevice(@Valid DeviceModel deviceModel) {
+    DeviceDTO deviceDto =
+        service
+            .findDevice(deviceModel.getSerialNumber())
+            .orElseThrow(
+                () -> new ExceptionType.DeviceNotFoundException(deviceModel.getSerialNumber()));
+
+    return Optional.of(deviceDto)
+        .map(service::save)
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(this::getUnknownException);
+  }
+
+  private RuntimeException getUnknownException() {
+    return new RuntimeException(UNKNOWN_ERROR_MESSAGE);
   }
```

Comment:
i think we can move this to the exception class

Reply:
I introduced a `UnknownException` for these. Good suggestion.